File: apps/android/app/src/main/java/online/akkdmsg/dakka/data/api/ChatActions.kt
package online.akkdmsg.dakka.data.api

import kotlinx.serialization.Serializable
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import online.akkdmsg.dakka.data.AppConfig
import online.akkdmsg.dakka.data.Chat
import online.akkdmsg.dakka.data.ChatMember
import online.akkdmsg.dakka.data.Message
import java.util.concurrent.TimeUnit

/** Действия над чатами не через сокет: создание, mute, clear, search, group admin. */
object ChatActions {

    private val http = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .build()

    // ── Создание чатов ──────────────────────────────────────────────────────

    @Serializable private data class DirectReq(val targetUserId: String)
    suspend fun createDirect(targetUserId: String): Chat =
        ApiClient.post("chats/direct", DirectReq(targetUserId))

    @Serializable private data class GroupReq(val name: String, val memberIds: List<String>)
    suspend fun createGroup(name: String, memberIds: List<String>): Chat =
        ApiClient.post("chats/group", GroupReq(name, memberIds))

    // ── Mute ────────────────────────────────────────────────────────────────

    @Serializable private data class MuteReq(val mutedUntil: String? = null)
    @Serializable private data class MuteResp(val mutedUntil: String? = null)

    suspend fun setMute(chatId: String, mutedUntil: String?) {
        ApiClient.post<MuteReq, MuteResp>(
            path = "chats/$chatId/mute",
            body = MuteReq(mutedUntil),
        )
    }

    // ── Clear history ───────────────────────────────────────────────────────

    @Serializable private data class ClearResp(val cleared: Boolean)
    suspend fun clearMessages(chatId: String): Boolean {
        val url = "${AppConfig.API_BASE_URL.trimEnd('/')}/chats/$chatId/messages".toHttpUrl()
        val req = Request.Builder().url(url).delete().apply {
            ApiClient.accessToken?.let { header("Authorization", "Bearer $it") }
            header("Accept", "application/json")
        }.build()
        val resp = http.newCall(req).execute()
        return resp.isSuccessful
    }

    // ── Search ──────────────────────────────────────────────────────────────

    suspend fun search(chatId: String, q: String): List<Message> =
        ApiClient.get("chats/$chatId/messages/search", query = mapOf("q" to q))

    // ── Group admin ─────────────────────────────────────────────────────────

    @Serializable private data class RenameReq(val name: String)
    suspend fun renameGroup(chatId: String, name: String): Chat {
        val url = "${AppConfig.API_BASE_URL.trimEnd('/')}/chats/$chatId".toHttpUrl()
        val bodyJson = kotlinx.serialization.json.Json.encodeToString(RenameReq.serializer(), RenameReq(name))
        val req = Request.Builder().url(url).patch(
            bodyJson.toRequestBody("application/json; charset=utf-8".toMediaType())
        ).apply {
            ApiClient.accessToken?.let { header("Authorization", "Bearer $it") }
            header("Accept", "application/json")
        }.build()
        val resp = http.newCall(req).execute()
        if (!resp.isSuccessful) throw RuntimeException("rename failed: ${resp.code}")
        return kotlinx.serialization.json.Json { ignoreUnknownKeys = true }.let { j ->
            val raw = resp.body?.string().orEmpty()
            val env = j.decodeFromString(
                online.akkdmsg.dakka.data.ApiEnvelope.serializer(Chat.serializer()),
                raw,
            )
            env.data ?: throw RuntimeException("empty rename response")
        }
    }

    @Serializable private data class AddMembersReq(val userIds: List<String>)
    @Serializable private data class AddMembersResp(val added: List<String>, val members: List<ChatMember>)
    suspend fun addGroupMembers(chatId: String, userIds: List<String>): List<ChatMember> {
        val resp: AddMembersResp = ApiClient.post(
            path = "chats/$chatId/members",
            body = AddMembersReq(userIds),
        )
        return resp.members
    }

    /** Универсально: kick (если userId не наш) или leave (если наш). */
    suspend fun removeMember(chatId: String, userId: String): Boolean {
        val url = "${AppConfig.API_BASE_URL.trimEnd('/')}/chats/$chatId/members/$userId".toHttpUrl()
        val req = Request.Builder().url(url).delete().apply {
            ApiClient.accessToken?.let { header("Authorization", "Bearer $it") }
            header("Accept", "application/json")
        }.build()
        val resp = http.newCall(req).execute()
        return resp.isSuccessful
    }
}

File: apps/ios/dakka/Core/Networking/ChatActions.swift
import Foundation

/// Тонкий слой над REST для действий над чатом, не идущих через сокет:
/// mute / clear history / search messages.
enum ChatActions {

    // MARK: - Mute

    @discardableResult
    static func setMute(chatId: String, mutedUntil: Date?) async throws -> Date? {
        struct Body: Codable { let mutedUntil: String? }
        struct Resp: Codable { let mutedUntil: Date? }
        let isoString = mutedUntil.map { ISO8601DateFormatter().string(from: $0) }
        let r: Resp = try await APIClient.shared.post(
            "chats/\(chatId)/mute",
            body: Body(mutedUntil: isoString)
        )
        return r.mutedUntil
    }

    // MARK: - Clear history

    @discardableResult
    static func clearMessages(chatId: String) async throws -> Bool {
        struct Resp: Codable { let cleared: Bool }
        var url = AppConfig.apiBaseURL
        url.append(path: "chats/\(chatId)/messages")
        var req = URLRequest(url: url)
        req.httpMethod = "DELETE"
        if let token = KeychainStore.get(KeychainStore.Keys.accessToken) {
            req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        let (data, response) = try await URLSession.shared.data(for: req)
        guard let http = response as? HTTPURLResponse, 200..<300 ~= http.statusCode else {
            throw URLError(.badServerResponse)
        }
        struct Env: Decodable { let success: Bool; let data: Resp? }
        let env = try JSONDecoder().decode(Env.self, from: data)
        return env.data?.cleared == true
    }

    // MARK: - Search

    static func search(chatId: String, query: String) async throws -> [Message] {
        try await APIClient.shared.get("chats/\(chatId)/messages/search", query: ["q": query])
    }

    // MARK: - Group admin

    /// PATCH /chats/:id — переименование/смена аватара.
    @discardableResult
    static func renameGroup(chatId: String, name: String) async throws -> Chat {
        struct Body: Codable { let name: String }
        var url = AppConfig.apiBaseURL
        url.append(path: "chats/\(chatId)")
        var req = URLRequest(url: url)
        req.httpMethod = "PATCH"
        if let token = KeychainStore.get(KeychainStore.Keys.accessToken) {
            req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = try JSONEncoder().encode(Body(name: name))
        let (data, _) = try await URLSession.shared.data(for: req)
        struct Env: Decodable { let data: Chat? }
        let env = try JSONDecoder.iso8601.decode(Env.self, from: data)
        if let chat = env.data { return chat }
        throw APIError.invalidResponse
    }

    /// POST /chats/:id/members
    static func addGroupMembers(chatId: String, userIds: [String]) async throws -> [ChatMember] {
        struct Body: Codable { let userIds: [String] }
        struct Resp: Codable { let added: [String]; let members: [ChatMember] }
        let r: Resp = try await APIClient.shared.post(
            "chats/\(chatId)/members",
            body: Body(userIds: userIds)
        )
        return r.members
    }

    /// DELETE /chats/:id/members/:userId — kick или leave (если userId === self)
    static func removeGroupMember(chatId: String, userId: String) async throws {
        var url = AppConfig.apiBaseURL
        url.append(path: "chats/\(chatId)/members/\(userId)")
        var req = URLRequest(url: url)
        req.httpMethod = "DELETE"
        if let token = KeychainStore.get(KeychainStore.Keys.accessToken) {
            req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        let (_, response) = try await URLSession.shared.data(for: req)
        guard let http = response as? HTTPURLResponse, 200..<300 ~= http.statusCode else {
            throw URLError(.badServerResponse)
        }
    }
}

extension JSONDecoder {
    static var iso8601: JSONDecoder {
        let d = JSONDecoder()
        d.dateDecodingStrategy = .iso8601
        return d
    }
}
